Userland/Libraries/LibWeb/HTML/HTMLImageElement.cpp

```diff
@@ -218,4 +219,536 @@ Optional<ARIA::Role> HTMLImageElement::default_role() const
+    return has_attribute(HTML::AttributeNames::srcset) || (parent() && is<HTMLPictureElement>(*parent()));
+}
+
+// https://html.spec.whatwg.org/multipage/images.html#update-the-image-data
+void HTMLImageElement::update_the_image_data(bool restart_animations)
+{
+    // 1. If the element's node document is not fully active, then:
+    if (!document().is_fully_active()) {
+        // FIXME: 1. Continue running this algorithm in parallel.
+        // FIXME: 2. Wait until the element's node document is fully active.
+        // FIXME: 3. If another instance of this algorithm for this img element was started after this instance
+        //           (even if it aborted and is no longer running), then return.
+        // FIXME: 4. Queue a microtask to continue this algorithm.
+    }
+
+    // 2. FIXME: If the user agent cannot support images, or its support for images has been disabled,
+    //           then abort the image request for the current request and the pending request,
+    //           set current request's state to unavailable, set pending request to null, and return.
+
+    // 3. Let selected source be null and selected pixel density be undefined.
+    Optional<String> selected_source;
+    Optional<float> selected_pixel_density;
+
+    // 4. If the element does not use srcset or picture
+    //    and it has a src attribute specified whose value is not the empty string,
+    //    then set selected source to the value of the element's src attribute
+    //    and set selected pixel density to 1.0.
+    if (!uses_srcset_or_picture() && has_attribute(HTML::AttributeNames::src) && !attribute(HTML::AttributeNames::src).is_empty()) {
+        selected_source = String::from_deprecated_string(attribute(HTML::AttributeNames::src)).release_value_but_fixme_should_propagate_errors();
+        selected_pixel_density = 1.0f;
+    }
+
+    // 5. Set the element's last selected source to selected source.
+    m_last_selected_source = selected_source;
+
+    // 6. If selected source is not null, then:
+    if (selected_source.has_value()) {
+        // 1. Parse selected source, relative to the element's node document.
+        //    If that is not successful, then abort this inner set of steps.
+        //    Otherwise, let urlString be the resulting URL string.
+        auto url_string = document().parse_url(selected_source.value().to_deprecated_string());
+        if (!url_string.is_valid())
+            goto after_step_6;
+
+        // 2. Let key be a tuple consisting of urlString, the img element's crossorigin attribute's mode,
+        //    and, if that mode is not No CORS, the node document's origin.
+        Optional<String> crossorigin_attribute;
+        if (auto crossorigin_value = attribute(HTML::AttributeNames::crossorigin); !crossorigin_value.is_null())
+            crossorigin_attribute = String::from_deprecated_string(crossorigin_value).release_value_but_fixme_should_propagate_errors();
```

Comment:
In `<media>` and `<script>`, I parsed/cached these in `parse_attribute` and `did_remove_attribute` so we only have to do it once (and avoids adding a null-DeprecatedString usage)